kcalcore/icaltimezones.cpp

```diff
@@ -1360,21 +1358,19 @@ QList<QDateTime> ICalTimeZoneSourcePrivate::parsePhase( icalcomponent *c,
         ICalFormat icf;
         ICalFormatImpl impl( &icf );
         impl.readRecurrence( icalproperty_get_rrule( p ), &r );
-        r.setStartDt( klocalStart );
-        // The end date time specified in an RRULE should be in UTC.
-        // Convert to local time to avoid timesInInterval() getting things wrong.
-        if ( r.duration() == 0 ) {
-          KDateTime end( r.endDt() );
-          if ( end.timeSpec() == KDateTime::Spec::UTC() ) {
-            end.setTimeSpec( KDateTime::Spec::ClockTime() );
-            r.setEndDt( end.addSecs( prevOffset ) );
-          }
+        r.setStartDt( kutcStart );
+
+        // the end date time specified must be in UTC
+        // we can not guarantee correctness if this is not the case
+        if (r.duration() == 0 && r.endDt().timeSpec() != KDateTime::Spec::UTC()) {
+            kDebug() << "UNTIL in RRULE must be specified in UTC";
+            return transitions;
```

Comment:
Minor details: inconsistent indentation. On error case and without debug enabled this silently ignores later defined rules. Ponder if such have kWarning() and either break for continuing to later rules or then return with empty transition list. Also misses qSortUnique() as done on general case. Then again this shouldn't happen so any behavior is not exactly wrong.